lib/spack/spack/repository.py

```diff
@@ -541,19 +541,24 @@ def repo_for_pkg(self, spec):
         """Given a spec, get the repository for its package."""
         # We don't @_autospec this function b/c it's called very frequently
         # and we want to avoid parsing str's into Specs unnecessarily.
+        namespace = None
         if isinstance(spec, spack.spec.Spec):
-            # If the spec already has a namespace, then return the
-            # corresponding repo if we know about it.
-            if spec.namespace:
-                fullspace = '%s.%s' % (self.super_namespace, spec.namespace)
-                if fullspace not in self.by_namespace:
-                    raise UnknownNamespaceError(spec.namespace)
-                return self.by_namespace[fullspace]
+            namespace = spec.namespace
             name = spec.name
-
         else:
             # handle strings directly for speed instead of @_autospec'ing
-            name = spec
+            try:
+                namespace, name = spec.split('.')
```

Comment:
Namespaces can be nested, so I think you want:

```
namespace, _, name = spec.rpartition('.')
```

to do this right.  This is what `Repo` uses.  `namespace` will be empty if there is no namespace.